geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/AvroSimpleFeatureTypeParser.scala

```diff
@@ -0,0 +1,411 @@
+      Option(builder.get(fieldName)).map(_.getUserData.putAll(metadata.extraProps))
+    }
+
+    val sft = builder.buildFeatureType()
+    sft.getUserData.putAll(sftUserData)
+    sft
+  }
+
+  @tailrec
+  private def addFieldToBuilder(builder: SimpleFeatureTypeBuilder,
+                                field: Schema.Field,
+                                typeOverride: Option[Schema.Type] = None): Unit = {
+    typeOverride.getOrElse(field.schema.getType) match {
+      case Schema.Type.STRING  => builder.add(field.name, classOf[java.lang.String])
+      case Schema.Type.BOOLEAN => builder.add(field.name, classOf[java.lang.Boolean])
+      case Schema.Type.INT     => builder.add(field.name, classOf[java.lang.Integer])
+      case Schema.Type.DOUBLE  => builder.add(field.name, classOf[java.lang.Double])
+      case Schema.Type.LONG    => builder.add(field.name, classOf[java.lang.Long])
+      case Schema.Type.FLOAT   => builder.add(field.name, classOf[java.lang.Float])
+      case Schema.Type.BYTES   => builder.add(field.name, classOf[Array[Byte]])
+      case Schema.Type.UNION   =>
+        // if a union has more than one non-null type, it is not supported
+        val types = field.schema.getTypes.map(_.getType).filter(_ != Schema.Type.NULL).toSet
+        if (types.size != 1) {
+          throw UnsupportedAvroTypeException(types.mkString("[", ", ", "]"))
+        } else {
+          addFieldToBuilder(builder, field, Option(types.head))
+        }
+      case Schema.Type.MAP     => throw UnsupportedAvroTypeException(Schema.Type.MAP.getName)
+      case Schema.Type.RECORD  => throw UnsupportedAvroTypeException(Schema.Type.RECORD.getName)
+      case Schema.Type.ENUM    => builder.add(field.name, classOf[java.lang.String])
+      case Schema.Type.ARRAY   => throw UnsupportedAvroTypeException(Schema.Type.ARRAY.getName)
+      case Schema.Type.FIXED   => throw UnsupportedAvroTypeException(Schema.Type.FIXED.getName)
+      case Schema.Type.NULL    => throw UnsupportedAvroTypeException(Schema.Type.NULL.getName)
+      case _                   => throw UnsupportedAvroTypeException("unknown")
+    }
+  }
+
+  private def parseMetadata(field: Schema.Field): GeoMesaAvroMetadata = {
+    // any field properties that are not one of the defined geomesa avro properties will go in the attribute user data
+    val extraProps = field.getProps.filterNot {
+      case (key, _) => reservedPropertyKeys.contains(key)
+    }.toMap
+
+    val exclude = GeoMesaAvroExcludeField.parse(field).getOrElse(false)
+
+    val geomFormat = GeoMesaAvroGeomFormat.parse(field)
+    val geomType = GeoMesaAvroGeomType.parse(field)
+    val geomDefault = GeoMesaAvroGeomDefault.parse(field).getOrElse(false)
+
+    if (geomFormat.isDefined && geomType.isDefined) {
+      return GeoMesaAvroMetadata(GeometryField(geomFormat.get, geomType.get, geomDefault), extraProps, exclude)
+    }
+
+    val dateFormat = GeoMesaAvroDateFormat.parse(field)
```

Comment:
I'm tempted to suggest changing these to lazy vals and then having if-elses to avoid the returns.  Not sure if that's faster/better or not.  Either way is probably fine.

@elahrvivaz any thoughts on that?

Reply:
Ah I see, I'll make that change